docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,175 @@
+```
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: guestbook
+spec:
+  generators:
+  - list:
+      elements:
+      - cluster: engineering-dev
+        url: https://1.2.3.4
+        env: dev
+      - cluster: engineering-prod
+        url: https://2.4.6.8
+        env: prod
+      - cluster: engineering-qa
+        url: https://9.8.7.6/
+        env: qa
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      steps:
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - dev
+          maxUpdate: 0 # if undefined or 0, all applications matched are updated together
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - qa
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - us-east-2
+                - eu-west-1
+                - ap-southeast-1
+          maxUpdate: 1 # maxUpdate supports both integer and percentage string values
+  template:
+    metadata:
+      name: '{{cluster}}-guestbook'
+      labels:
+        env: "{{env}}"                                # label can be provided explicitly from a list generator
+        region: "{{metadata.labels.cluster/region}}"  # or pulled from labels on the argo cluster secrets
+    spec:
+      source:
+        repoURL: https://github.com/infra-team/cluster-deployments.git
+        targetRevision: HEAD
+        path: guestbook/{{cluster}}
+      destination:
+        server: '{{url}}'
+        namespace: guestbook
+```
+
+In the above example, when the guestbook ApplicationSet is created or modified, the Application resources are each updated in the order defined in `strategy.rollingUpdate`. In this case, all generated Applications (applied or not) with a label that matches the expression `env: dev` are updated to match the template. All Applications in this step are updated in parallel, because the `maxUpdate` is set to zero. The rolling update strategy progresses after the first set of Applications has successfully progressed and become healthy again. Progress towards the next step starts only after the current step has completely finished, regardless of the `maxUpdate` value. The `maxUpdate` field only throttles the total number of matching Applications updating in the current step. After the first step completes, the ApplicationSet updates all Application resources with label `env: qa` at the same time, because `maxUpdate` is undefined. Finally, during the third step, the Application resources labeled `region: us-east-2`, `region: eu-west-1`, or `region: ap-southeast-1` are updated, one by one, as the `maxUpdate` for the final step is 1.
+
+An Application rollout is considered “complete” when the Application resource has been:
+- Synced successfully.
+- Moved into a “Progressing” state.
+- Moved out of a “Progressing” state and into a “Healthy” state.
```

Comment:
I've actually seen another detrimental behavior that may or may not be something this can deal with. 

With extremely dynamic clusters, we have many apps that for long periods of time never exit the progressing state because they contain something like a daemonset that is always adding or removing nodes. This has bitten us pretty hard on Sync Waves and hadn't found any way around it.

Reply:
We typically see the same behavior running Datadog as a daemonset on our older clusters we're working to decommission. I think the currently proposal would still hit the same problem, and the ApplicationSet rollout would effectively be blocked on any Application resources that are stuck in a Progressing state due to any pod controller taking a long time to become Ready.

I don't think it would be difficult to add a configurable Readiness timeout in the future if that's a desirable use case for some ApplicationSet rollouts.